crypto/cipher_extra/aead_test.cc

```diff
@@ -271,6 +271,203 @@ TEST_P(PerAEADTest, TestVector) {
+  // Below ctrl returns hmac|pad len for encryption.
+  // For decryption, it returns digest size.
+  EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_TLS1_AAD, EVP_AEAD_TLS1_AAD_LEN, ad);
+}
+
+// |EVP_aes_128/256_cbc_hmac_sha1/256| tests.
+// Tests a TLS specific legacy AEAD against a series of test vectors from a file, using the
+// FileTest format. As an example, here's a valid test case:
+//
+// # DIGEST: 918c0df73de553b5bdffb7365f93a430292f6eea
+// KEY: 2993a340b9b3c589c7481df3f4183aa23fd8d7efd88503f78b8ed1c8e9ba2fd6773e0d0c
+// NONCE: 302a5f47e037446f5891d77df660ed82
+// IN: 302a5f47e037446f5891d77df660ed8286d641b877
+// AD: 97b684e0fb56f97c3903020015
+// CT: 6854710b76c79c9be84b3669fc4f05b17e67a9cc14
+// TAG: 9c6998bf3b172670c5f8613f479641468bbed4c68d093940c92de4
+// TAG_LEN: 20
+TEST_P(PerTLSLegacyAEADTest, TestVector) {
+  const EVP_CIPHER *cipher = legacy_aead();
+  if (!cipher) {
+    // This condition can happen when
+    // 1. !defined(AES_CBC_HMAC_SHA_STITCH).
+    // 2. |hwaes_capable| returns false.
+    //    -- Checked by Intel SDE. CPU is "mrm", "nhm", "pnr", "slt" and "p4p".
+    return;
+  }
+  std::string test_vectors = "crypto/cipher_extra/test/";
+  test_vectors += GetParam().test_vectors;
+  FileTestGTest(test_vectors.c_str(), [&](FileTest *t) {
+    std::vector<uint8_t> key, nonce, in, ad, ct, tag;
+    ASSERT_TRUE(t->GetBytes(&key, "KEY"));
+    ASSERT_TRUE(t->GetBytes(&nonce, "NONCE"));
+    ASSERT_TRUE(t->GetBytes(&in, "IN"));
+    ASSERT_TRUE(t->GetBytes(&ad, "AD"));
+    ASSERT_TRUE(t->GetBytes(&ct, "CT"));
+    ASSERT_TRUE(t->GetBytes(&tag, "TAG"));
+    size_t tag_len = tag.size();
+    if (t->HasAttribute("TAG_LEN")) {
+      // Legacy AEADs are MAC-then-encrypt and may include padding in the TAG
+      // field. TAG_LEN contains the actual size of the digest in that case.
+      std::string tag_len_str;
+      ASSERT_TRUE(t->GetAttribute(&tag_len_str, "TAG_LEN"));
+      tag_len = strtoul(tag_len_str.c_str(), nullptr, 10);
+      ASSERT_TRUE(tag_len);
+    }
+    bool explicit_iv = !nonce.empty();
+    size_t aes_block_size = EVP_CIPHER_block_size(cipher);;
+    size_t key_block_size = EVP_CIPHER_key_length(cipher);
+    size_t e_iv_len = 0;
+    bssl::ScopedEVP_CIPHER_CTX ctx;
+
+    std::vector<uint8_t> encrypted(ct.size() + tag.size());
+    // The |key| is Mac key + AES key + IV.
+    size_t mac_key_size = tag_len;
+    const uint8_t *aes_key = key.data() + mac_key_size;
+    std::vector<uint8_t> iv(aes_block_size);
+    if (explicit_iv) {
+      e_iv_len = aes_block_size;
+      OPENSSL_memcpy(iv.data(), nonce.data(), nonce.size());
+    } else {
+      OPENSSL_memcpy(iv.data(), key.data() + mac_key_size + key_block_size, aes_block_size);
+    }
+    if (!t->HasAttribute("NO_SEAL") &&
+        !(GetParam().flags & kNondeterministic)) {
+      // Even without |EVP_CIPHER_CTX_set_padding|, |EVP_Cipher| returns error code because
```

Comment:
Not sure I get this comment?

Reply:
`EVP_CIPHER_CTX_set_padding` is used to set a flag so other cipher APIs will perform padding if the input size is not integral multiple of the cipher block size. The imported API ignores the flag even when it's set. It requires the input `len` to be integral multiple of `AES_BLOCK_SIZE`.

s2n code uses above API. I copied the usage here to tell `no automatic padding`.

Reply:
stitched can only be used through `EVP_Cipher()` and not through the streaming APIs. So, I think I agree that setting this flag doesn't actually matter. But we can leave it.